### PR TITLE
Add /library option

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes
Introduce the command line option `/library:<file>`:
```
/library:<file>
    The contents of this file and any files it includes can be referenced from other files as if they were included. 
    However, these contents are skipped during code generation and verification.
    This option is useful in a diamond dependency situation, 
    to prevent code from the bottom dependency from being generated more than once.
```

I have not updated the Dafny reference manual intentionally, since we're expecting to deprecate this option once we add support for project files and the ability to configure dependencies there.

### Tests
- Add a lit test

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: docs/DafnyRef/Options.txt [hunk 1]
```diff
@@ -361,6 +361,9 @@ Usage: Dafny [ option ... ] [ filename ... ]
   /useRuntimeLib
       Refer to pre-built DafnyRuntime.dll in compiled assembly rather
       than including DafnyRuntime.cs verbatim.
+  /library:<file>
+      The contents of this file and any files it includes will be included in any generated code,
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply docs/DafnyRef/Options.txt]: Ah, updated this outdated file using `make refman`

File: Test/comp/Libraries/consumer.dfy [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+// CHECK: GloballyUniqueProducer
+
+include "Inputs/indirectLibrary.dfy"
+include "Inputs/directLibrary.dfy"
+
+module ConsumingModule {
+  import A = GloballyUniqueProducer.ExportingModule
+  import B = GloballyUniqueProducer.ExportingModule
+
+  const myConstant := A.exportedVariable + A.exportedVariable
```
Comment (cpitclaudel) [inline Test/comp/Libraries/consumer.dfy:15]:
```suggestion
  const myConstant := A.exportedVariable + B.exportedVariable
```
?

**Comment by cpitclaudel** [comment]
To add to this: we "already" have support for this, if the code comes from a DLL; wouldn't it be enough to add a flag so that all includes are treated as DLLs?

**Comment by keyboardDrummer** [comment]
> To take advantage of this new options, do users have to use `/library` for each file that is part of a library that they want to use?  I was thinking that we would support "separate compilation" instead by adding a flag to spill a each file separately without their includes, so that to compile a project you would call `Dafny /spillTargetCode:OnlyTheseFiles /noVerify a.dfy b.dfy c.dfy …`

They can specify a single root library file that includes all other library files.

**Comment by cpitclaudel** [comment]
> In the short term, I think either what you suggest or the option in this PR work well enough. Note that with the /library option, users don't need to use includes, because the file passed to /library and files it includes, are included by default. 

I'm not comfortable with this change, since AFAICT it will make it more complicated to edit files in the IDE. Do we even have a `/library` flag for the language server?  In Emacs it won't be an issue because the legacy server accepts arbitrary flags, but it sounds like it *will* be an issue for VSCode, no?

I imagine people just won't use this feature, but then why have it? If course this goes away once we have project files.

> The continuation of this option is to enable specifying the libraries in a Dafny project file.

Indeed, this will be nice.

> In the long term, I was planning on removing the need for includes entirely. 

That's nice.  But this will be orthogonal to `library`, since we'll want to "auto-include" also files that are not part of an external library, right?

> Were you thinking of specifying dependencies inline in Dafny files? Like include https://github.com/dafny-lang/libraries/blob/44f5891e7b5b21a5f13f139b22fb5559b46de36f/src/BoundedInts.dfy ?

No, that sounds very scary.

**Review by cpitclaudel — APPROVED** [review]

**Comment by cpitclaudel** [comment]
For the record, I will also add another comment:  I would prefer a model in which we have an explicit include list, rather than an explicit exclude list. This is the usual model with "object files", where compiling a C file with includes compiles the `.c` file to a `.o`, or compiling an OCaml `.ml` file produces `.cmo` / `.cma` / `.cmxa` files, etc.

With such a feature, one could simply compile all files separately, with individual calls to `Dafny -compile`, and the help of a makefile.  This low-level feature would then be wrapped in higher-level logic in our project-management code.

This is in part because IME, one often has to use a Makefile around large Dafny projects anyway, because of `assume` not being compiled.

**Comment by keyboardDrummer** [comment]
> I'm not comfortable with this change, since AFAICT it will make it more complicated to edit files in the IDE. Do we even have a `/library` flag for the language server? In Emacs it won't be an issue because the legacy server accepts arbitrary flags, but it sounds like it _will_ be an issue for VSCode, no?

Users should keep their includes clauses, so the IDE will still work the same. The only difference AFAICT between the `/library` flag and the `/spillTargetCode:OnlyTheseFiles` is then respectively whether the root source file and root library file need to be mentioned in the CLI or all source files but non of the library files. I suggest we use the /library option since it's closer to how project files will work, where the project file will declare libraries.

> For the record, I will also add another comment: I would prefer a model in which we have an explicit include list, rather than an explicit exclude list. This is the usual model with "object files", where compiling a C file with includes compiles the .c file to a .o, or compiling an OCaml .ml file produces .cmo / .cma / .cmxa files, etc.

We can add "include" and "exclude" fields to the project files that allow users to setup which .dfy files are considered source. 

> With such a feature, one could simply compile all files separately, with individual calls to Dafny -compile, and the help of a makefile. This low-level feature would then be wrapped in higher-level logic in our project-management code.

Wouldn't compiling all files separately cause any file that is used by another will need to be parsed and resolved more than once?

> That's nice. But this will be orthogonal to library, since we'll want to "auto-include" also files that are not part of an external library, right?

I wouldn't say that. You could translate the project file definition to a CLI call where /library is used and all source files configured in the project file are passed a source files.

**Comment by cpitclaudel** [comment]
> Users should keep their includes clauses, so the IDE will still work the same.

Got it, thanks.  But then why do we allow them to remove them in this PR?

> The only difference AFAICT between the /library flag and the /spillTargetCode:OnlyTheseFiles is then respectively whether the root source file and root library file need to be mentioned in the CLI or all source files but non of the library files. I don't have a strong preference for either.

I have a strong preference for mentioning "this project"'s files, because that's how things work in other compilers that I'm familiar with.

> Wouldn't compiling all files separately cause any file that is used by another will need to be parsed and resolved more than once?

That's right.  I suspect this wouldn't be an issue, because it's the translation to Boogie and to Z3 that's slow, not our own typechecking.

We already have (mostly) the behavior that I like best with the Java compiler, because it has to have one class per file.  The only additional aspect there is respecting the list of files passed on the CLI.

**Comment by keyboardDrummer** [comment]
> I have a strong preference for mentioning "this project"'s files, because that's how things work in other compilers that I'm familiar with.

I don't understand how that can work without requiring include clauses in your source files, which I want to avoid. Indeed this PR doesn't avoid the need for includes clauses yet because you need them for IDE support, but we'll get to update the IDE so you won't need them and this PR works towards that.

When include clauses are not used, both this PR's approach and what you suggest will require defining all source files on the CLI. Even if include clauses are used, I would suggest users define all their source files on the CLI to ensure that Dafny will verify them.

> > Users should keep their includes clauses, so the IDE will still work the same.
> 
> Got it, thanks. But then why do we allow them to remove them in this PR?

I'm not sure what you mean by allow, but for IDE support, users should keep the includes clauses after this PR. Sorry for the confusion!

> > Wouldn't compiling all files separately cause any file that is used by another will need to be parsed and resolved more than once?
> 
> That's right. I suspect this wouldn't be an issue, because it's the translation to Boogie and to Z3 that's slow, not our own typechecking.

Note that it's a slowness that scales with the include depth of a program. For a program that's a chain of files that include each other, the resolving time would increase quadratically with the chain size (keeping the amount of code in each link the same)

**Pull request merged**